tests/common/exceptions.h

```diff
@@ -0,0 +1,302 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+*******************************************************************************/
+
+#ifndef __SYCLCTS_TESTS_COMMON_EXCEPTIONS_H
+#define __SYCLCTS_TESTS_COMMON_EXCEPTIONS_H
+
+#include <catch2/catch_tostring.hpp>
+#include <catch2/catch_translate_exception.hpp>
+#include <catch2/matchers/catch_matchers_templated.hpp>
+#include <sycl/sycl.hpp>
+
+#include "../../util/logger.h"
+#include "../common/conversion.h"
+#include "../common/macros.h"
+
+#include <sstream>
+#include <string>
+#include <utility>
+
+CATCH_REGISTER_ENUM(sycl::errc, sycl::errc::success, sycl::errc::runtime,
+                    sycl::errc::kernel, sycl::errc::accessor,
+                    sycl::errc::nd_range, sycl::errc::event,
+                    sycl::errc::kernel_argument, sycl::errc::build,
+                    sycl::errc::invalid, sycl::errc::memory_allocation,
+                    sycl::errc::platform, sycl::errc::profiling,
+                    sycl::errc::feature_not_supported,
+                    sycl::errc::kernel_not_supported,
+                    sycl::errc::backend_mismatch);
+
+namespace sycl_cts {
+namespace detail {
+namespace has_member {
+
+// Safely ensures class member availability w/o compile-time failures using
+// SFINAE
+template <typename, typename = void>
+struct code : std::false_type {};
+template <typename T>
+struct code<T, std::void_t<decltype(std::declval<T>().code())>>
+    : std::true_type {};
+
+template <typename, typename = void>
+struct category : std::false_type {};
+template <typename T>
+struct category<T, std::void_t<decltype(std::declval<T>().category())>>
+    : std::true_type {};
+
+}  // namespace has_member
+
+namespace has_function {
+
+// Safely ensures free function availability w/o compile-time failures
+template <typename = void>
+struct sycl_make_error_code : std::false_type {};
+template <>
+struct sycl_make_error_code<
+    std::void_t<decltype(sycl::make_error_code(std::declval<sycl::errc>()))>>
+    : std::true_type {};
+
+}  // namespace has_function
+}  // namespace detail
```

Comment:
Correct me if I'm wrong, but these SFINAE helpers test for *non-optional* features in implementations that haven't quite caught up yet, right? If so, I'd suggest to either add a note that these should be removed at some point, or preferably in view, simply do something like

```c++
#if !defined(__COMPUTECPP__) && !defined(__HIPSYCL__)
constexpr bool has_member_code = true;
constexpr bool has_member_category = true;
#else
constexpr bool has_member_code = false;
constexpr bool has_member_category = false;
#endif
```

Reply:
The point here is that such header would be used within all SYCL CTS tests, so it's not intended to be removed.
I assume there would be two possible use-cases here.
 - Any new or existing SYCL implementation doesn't support non-optional API for SYCL exceptions
 - Any existing implementation encounter compilation error for already supported non-optional API

In any of these cases:
 - we could still gain from "what()" messages at least
 - it would be better to have specific SYCL CTS tests failed rather then all of them

For example:
 - there should be no failure if the test just includes `exceptions.h` or `common.h`
 - test should not fail if it just prints out exception information
 - test should fail on specific CATCH_THROW_MATCHES in case specific check is unavailable

Are there any concerns regarding suggested solution?

Reply:
To be clear, I'm not questioning the intent of this code. I agree that it is important that fundamental CTS infrastructure such as this ensures compatibility with all implementations. However I don't think this should be a permament mechanism, and I'm not convinced that the complexity of introducing SFINAE helpers here is warranted either.

> so it's not intended to be removed.

All three SYCL implementations currently supported by the CTS are striving to become conformant at some point, which is when we could remove these helpers.

> Any new [...] SYCL implementation doesn't support non-optional API for SYCL exceptions

I don't think we should optimize for this hypothetical case. If we wanted any new SYCL implementation to immediately be able to compile the CTS, we would have to have these kinds of SFINAE helpers for *any and all* SYCL API interactions in CTS infrastructure code, which clearly we don't (and realistically, can't) have.

---

So all I'm saying is this: Ensuring compatibility with CCPP, DPC++ and hipSYCL is good (and required for CI), so having something like this (or my proposed `constexpr bool`s) is required at the moment. However I believe we should get rid of it as soon as those three implementations no longer require the workaround, and so we don't forget, I'm suggesting you add a TODO note :).

Or am I missing something?

Reply:
Agree) Updated the code with feature flags and TODO notes regarding feature flags removal.

Preprocessor definitions used instead of constexpr variables because:
- looks like hipSYCL doesn't support `sycl::errc` enum yet
- there are too many cases when discarded constexpr substatement could still result in compilation failure